pkg/cli/run.go

```diff
@@ -27,87 +27,185 @@ import (
+	UseColors bool
+	useColors string // used for flag processing
+
+	// Verbose indicates whether to display all fields in default human readlbe output
+	Verbose bool
+}
+
+// Scan executes the terrascan scan command
+func (s *ScanOptions) Scan() error {
+	if err := s.Init(); err != nil {
+		zap.S().Error("scan init failed", zap.Error(err))
+		return err
+	}
+
+	if err := s.Run(); err != nil {
+		zap.S().Error("scan run failed", zap.Error(err))
+		return err
+	}
+	return nil
+}
+
+//Init initalises and validates ScanOptions
+func (s *ScanOptions) Init() error {
+	s.initColor()
+	err := s.validate()
+	if err != nil {
+		zap.S().Error("failed to start scan", zap.Error(err))
+		return err
+	}
+	return nil
+}
+
+// validate config only for human readable output
+// rest command options are validated by the executor
+func (s ScanOptions) validate() error {
+	// human readable output doesn't support --config-only flag
+	// if --config-only flag is set, then exit with an error
+	// asking the user to use yaml or json output format
+	if s.configOnly && strings.EqualFold(s.outputType, humanOutputFormat) {
+		return errors.New("please use yaml or json output format when using --config-only flag")
+	}
+	return nil
+}
+
+// initialises use colors options
+func (s *ScanOptions) initColor() {
+	switch strings.ToLower(s.useColors) {
+	case "auto":
+		if isatty.IsTerminal(os.Stdout.Fd()) || isatty.IsCygwinTerminal(os.Stdout.Fd()) {
+			s.UseColors = true
+		} else {
+			s.UseColors = false
+		}
+
+	case "true":
+		fallthrough
+	case "t":
+		fallthrough
+	case "y":
+		fallthrough
+	case "1":
+		fallthrough
+	case "force":
+		s.UseColors = true
+
+	default:
+		s.UseColors = false
+	}
+}
+
 // Run executes terrascan in CLI mode
-func Run(iacType, iacVersion string, cloudType []string,
-	iacFilePath, iacDirPath, configFile string, policyPath []string,
-	format, remoteType, remoteURL string, configOnly, useColors, verbose bool) {
+func (s *ScanOptions) Run() error {
 
 	// temp dir to download the remote repo
 	tempDir := filepath.Join(os.TempDir(), utils.GenRandomString(6))
 	defer os.RemoveAll(tempDir)
 
 	// download remote repository
-	path, err := downloadRemoteRepository(remoteType, remoteURL, tempDir)
+	err := s.downloadRemoteRepository(tempDir)
 	if err != nil {
-		return
-	}
-
-	if path != "" {
-		iacDirPath = path
+		return err
 	}
 
 	// create a new runtime executor for processing IaC
-	executor, err := runtime.NewExecutor(iacType, iacVersion, cloudType,
-		iacFilePath, iacDirPath, configFile, policyPath)
+	executor, err := runtime.NewExecutor(s.iacType, s.iacVersion, s.policyType,
+		s.iacFilePath, s.iacDirPath, s.configFile, s.policyPath)
 	if err != nil {
-		return
+		return err
 	}
 
 	// executor output
 	results, err := executor.Execute()
 	if err != nil {
-		return
+		return err
 	}
 
 	// write results to console
-	err = writeResults(results, useColors, verbose, configOnly, format)
+	err = s.writeResults(results)
```

Comment:
Same comment for writeResults(), it feels weird to add as part of ScanOptions... Perhaps instead, we should have a separate Scanner struct that accepts scan options and performs these tasks... The Scanner object overall could be the thing that orchestrates operations for scan options, policies, normalized input, the scan engine, and results output

Reply:
I'll add that my own previous refactor took the policy input directly into the policy engine, but something felt missing there as well. But having an overall Scanner class that handles this looks like a good option.

Reply:
Had a discussion about this w/Pankaj/Yusuf; I meant for the main Scanner object to be within the Executor(), not outside, so since this change is all outside of the Executor, it's not applicable, please disregard.